From 526674a52ef9171741c30c566e07796819890dfd Mon Sep 17 00:00:00 2001
From: Refael Ackermann <refack@gmail.com>
Date: Fri, 17 Mar 2017 14:05:47 -0400
Subject: [PATCH 4/4] win: Make buildable on vs2017
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

changes vcbuild.bat arg from vc2015 to vs2015/vs2017
`vc` as in Visual C++ is actually versions 14.0 or 14.10
`vs` as in Visual Studio is 2015 or 2017
ref: http://lists.boost.org/Archives/boost/2017/03/233597.php ðŸ¤¦
---
 BUILDING.md |  4 +++-
 configure   | 12 ++++++++++++
 vcbuild.bat | 29 ++++++++++++++++++++++++++---
 3 files changed, 41 insertions(+), 4 deletions(-)

diff --git BUILDING.md BUILDING.md
index b325972f4b..64ed4087e3 100644
--- BUILDING.md
+++ BUILDING.md
@@ -38,7 +38,7 @@ Support is divided into three tiers:
 |--------------|--------------|----------------------------------|----------------------|------------------|
 | GNU/Linux    | Tier 1       | kernel >= 2.6.18, glibc >= 2.5   | x86, x64, arm, arm64 |                  |
 | macOS        | Tier 1       | >= 10.10                         | x64                  |                  |
-| Windows      | Tier 1       | >= Windows 7 or >= Windows2008R2 | x86, x64             |                  |
+| Windows      | Tier 1       | >= Windows 7 / 2008 R2           | x86, x64             | vs2015 or vs2017 |
 | SmartOS      | Tier 2       | >= 15 < 16.4                     | x86, x64             | see note1        |
 | FreeBSD      | Tier 2       | >= 10                            | x64                  |                  |
 | GNU/Linux    | Tier 2       | kernel >= 4.2.0, glibc >= 2.19   | ppc64be              |                  |
@@ -177,6 +177,8 @@ Prerequisites:
   * [Visual Studio 2015 Update 3](https://www.visualstudio.com/), all editions
     including the Community edition (remember to select
     "Common Tools for Visual C++ 2015" feature during installation).
+  * [Visual Studio 2017](https://www.visualstudio.com/downloads/), any edition (including the Build Tools SKU).
+    __Required Components:__ "MSbuild", "VC++ 2017 v141 toolset" and one of the Windows SDKs (10 or 8.1).
 * Basic Unix tools required for some tests,
   [Git for Windows](http://git-scm.com/download/win) includes Git Bash
   and tools which can be included in the global `PATH`.
diff --git configure configure
index 31d682575e..8da9d10594 100755
--- configure
+++ configure
@@ -995,6 +995,17 @@ def configure_static(o):
       if options.enable_asan:
         o['libraries'] += ['-static-libasan']
 
+def config_vs2017(o):
+  if os.environ.get('VisualStudioVersion') != '15.0': return
+  try:
+    o['msbuild_toolset'] = 'v141'
+    sdk_ver = os.environ['WindowsSDKVersion'].rstrip("\\")
+    o['msvs_windows_target_platform_version'] = sdk_ver
+  except Exception as e:
+    print(' Error: could not find a compatible SDK, ' +
+          'make sure to run from a "VS developer command prompt"')
+    sys.exit(1)
+
 
 def write(filename, data):
   filename = os.path.join(root_dir, filename)
@@ -1316,6 +1327,7 @@ configure_openssl(output)
 configure_intl(output)
 configure_static(output)
 configure_inspector(output)
+config_vs2017(output)
 
 # variables should be a root level element,
 # move everything else to target_defaults
diff --git vcbuild.bat vcbuild.bat
index 9cbac9088f..d6e63a32f8 100644
--- vcbuild.bat
+++ vcbuild.bat
@@ -1,4 +1,4 @@
-@echo off
+@IF NOT DEFINED DEBUG_VCBUILD echo off
 
 cd %~dp0
 
@@ -50,7 +50,8 @@ if /i "%1"=="clean"         set target=Clean&goto arg-ok
 if /i "%1"=="ia32"          set target_arch=x86&goto arg-ok
 if /i "%1"=="x86"           set target_arch=x86&goto arg-ok
 if /i "%1"=="x64"           set target_arch=x64&goto arg-ok
-if /i "%1"=="vc2015"        set target_env=vc2015&goto arg-ok
+if /i "%1"=="vs2015"        set target_env=vs2015&goto arg-ok
+if /i "%1"=="vs2017"        set target_env=vs2017&goto arg-ok
 if /i "%1"=="noprojgen"     set noprojgen=1&goto arg-ok
 if /i "%1"=="nobuild"       set nobuild=1&goto arg-ok
 if /i "%1"=="nosign"        set "sign="&echo Note: vcbuild no longer signs by default. "nosign" is redundant.&goto arg-ok
@@ -140,7 +141,29 @@ if defined noprojgen if defined nobuild if not defined sign if not defined msi g
 
 @rem Set environment for msbuild
 
+@rem Look for Visual Studio 2017
+:vs-set-2017
+if defined target_env if "%target_env%" NEQ "vs2017" goto vs-set-2015
+if defined VisualStudioVersion if "%VisualStudioVersion%"=="15.0" goto found_vs2017
+echo Looking for Visual Studio 2017
+set cl141cmd="%~dp0tools\vc141helper\get_key_helper.cmd"
+for /F "tokens=*" %%A IN ('cmd /D /S /C "%cl141cmd% IsVcCompatible InstallationPath"') DO set VS2017_INST_PATH=%%A
+if "_%VS2017_INST_PATH%_" == "__" goto vs-set-2015
+if NOT EXIST "%VS2017_INST_PATH%" goto vs-set-2015
+@rem GYP_MSVS_VERSION=2015 is a workaround for old `GYP`
+if %target_arch%==x64 SET VS_TARGET_ARCH=amd64
+set "VSVARSALL="%VS2017_INST_PATH%\VC\Auxiliary\Build\vcvarsall.bat" %VS_TARGET_ARCH%"
+call %VSVARSALL%
+:found_vs2017
+echo Found MSVS version %VisualStudioVersion%
+set GYP_MSVS_VERSION=2015
+set PLATFORM_TOOLSET=v141
+set VS_TARGET_ARCH=x86
+goto msbuild-found
+
 @rem Look for Visual Studio 2015
+:vs-set-2015
+if defined target_env if "%target_env%" NEQ "vs2015" goto msbuild-not-found
 echo Looking for Visual Studio 2015
 if not defined VS140COMNTOOLS goto msbuild-not-found
 if not exist "%VS140COMNTOOLS%\..\..\vc\vcvarsall.bat" goto msbuild-not-found
@@ -417,7 +440,7 @@ echo Failed to create vc project files.
 goto exit
 
 :help
-echo vcbuild.bat [debug/release] [msi] [test-all/test-uv/test-inspector/test-internet/test-pummel/test-simple/test-message] [clean] [noprojgen] [small-icu/full-icu/without-intl] [nobuild] [sign] [x86/x64] [vc2015] [download-all] [enable-vtune] [lint/lint-ci]
+echo vcbuild.bat [debug/release] [msi] [test-all/test-uv/test-inspector/test-internet/test-pummel/test-simple/test-message] [clean] [noprojgen] [small-icu/full-icu/without-intl] [nobuild] [sign] [x86/x64] [vs2015/vs2017] [download-all] [enable-vtune] [lint/lint-ci]
 echo Examples:
 echo   vcbuild.bat                : builds release build
 echo   vcbuild.bat debug          : builds debug build
-- 
2.12.2.windows.1

